store/cachekv/internal/memcache.go

```diff
@@ -0,0 +1,93 @@
+package internal
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/tidwall/btree"
+)
+
+var errKeyEmpty = errors.New("key cannot be empty")
+
+const (
+	// The approximate number of items and children per B-tree node. Tuned with benchmarks.
+	// copied from memdb.
+	bTreeDegree = 32
+)
+
+// MemCache implements the in-memory cache for cachekv store,
+// we don't use MemDB here because cachekv is used extensively in sdk core path,
+// we need it to be as fast as possible, while `MemDB` is mainly used as a mocking db in unit tests.
+//
+// We choose tidwall/btree over google/btree here because it provides API to implement step iterator directly.
+type MemCache struct {
+	tree *btree.BTreeG[item]
+}
+
+// NewMemCache creates a wrapper around `btree.BTreeG`.
+func NewMemCache() MemCache {
+	return MemCache{tree: btree.NewBTreeGOptions(byKeys, btree.Options{
+		Degree:  bTreeDegree,
+		NoLocks: false,
+	})}
+}
+
+// Set set a cache entry, dirty means it's newly set,
+// `nil` value means a deletion.
+func (bt MemCache) Set(key, value []byte, dirty bool) {
+	bt.tree.Set(item{key: key, value: value, dirty: dirty})
+}
+
+// Get returns (value, found)
+func (bt MemCache) Get(key []byte) ([]byte, bool) {
+	i, found := bt.tree.Get(item{key: key})
+	if !found {
+		return nil, false
+	}
+	return i.value, true
+}
+
+// ScanDirtyItems iterate over the dirty entries.
+func (bt MemCache) ScanDirtyItems(fn func(key, value []byte)) {
+	bt.tree.Scan(func(item item) bool {
+		if item.dirty {
+			fn(item.key, item.value)
+		}
+		return true
+	})
+}
+
+// Copy the cache. This is a copy-on-write operation and is very fast because
+// it only performs a shadowed copy.
```

Comment:
Can you please explain a bit how this work? I understand that the copy is not done until the btree is modified. This means either the original or the copy?

Reply:
https://github.com/tidwall/btree/blob/master/btreeg.go#L296
I think the writer will check a seq number on nodes and do the copy when needed.